pipeline/detection.py

```diff
@@ -706,6 +757,121 @@ def extract_sources_sep(self, image):
 
         return sources
 
+    def extract_sources_filter(self, image):
+        """Find sources in an image using the matched-filter method.
+
+        If the image has a "score" array, will use that.
+        If not, will apply the PSF of the image to make a score array.
+
+        Sources are detected as points in the score image that are
+        above the given threshold. If points are too close, they
+        will be merged into a single source.
+        The output SourceList will include the aperture and PSF photometry,
+        and the x,y positions will be given using the source centroid.
+
+        # TODO: should we do iterative PSF tapered centroiding?
+
+        Parameters
+        ----------
+        image: Image
+            The image to extract sources from. Must have a score, or a PSF,
+            that can be gotten from one of these places (in order):
+            - a PSF object loaded on the Image object itself.
+            - a zogy_psf attribute attached to the Image object.
+            - a PSF object of the new_image attribute of the Image object.
+            If the image also has a zogy_alpha attribute, it will be used
+            to extract the PSF photometry.
+
+        Returns
+        -------
+        sources: SourceList
+            The list of sources detected in the image.
+            This contains a table where each row represents
+            one source that was detected, along with all its properties.
+
+        """
+        score = image.score.copy()
+
+        if score is None:
+            raise NotImplementedError('Still need to add the matched-filter cross correlation! ')
+
+        psf_image = None
+        if image.psf is not None:
+            psf_image = image.psf.get_clip()
+        elif getattr(image, 'zogy_psf', None) is not None:
+            psf_image = image.zogy_psf
+
+        fwhm = image.fwhm_estimate
+        if fwhm is None:
+            fwhm = image.new_image.fwhm_estimate
+        if fwhm is None and image.psf is not None:
+            fwhm = image.psf.fwhm_pixels
+        if fwhm is None and image.new_image.psf is not None:
+            fwhm = image.new_image.psf.fwhm_pixels
+
+        if fwhm is None:
+            raise RuntimeError("Cannot find a FWHM estimate from the given image or its new_image attribute.")
+
+        # typical scale of PSF is X times the FWHM
+        fwhm_pixels = max(int(np.ceil(fwhm * self.pars.separation_fwhms / image.instrument_object.pixel_scale)), 1)
+        # remove flagged pixels
+        score[image.flags > 0] = np.nan
+
+        # remove the edges
+        border = fwhm_pixels * 2
+        score[:border, :] = np.nan
+        score[-border:, :] = np.nan
+        score[:, :border] = np.nan
+        score[:, -border:] = np.nan
+
+        # normalize the score based on sigma_clipping
+        # TODO: we should check if we still need this after b/g subtraction on the input images
```

Comment:
(Is there an issue?)

Reply:
Yeah, there are a few issues on background subtraction, I'm going to deal with this after Cutouts and Measurements.